File: src/Solutions/Euler031.hs
module Solutions.Euler031 where
import Utils.Operators((<:>))
import Data.Array

coins = listArray (0, 7) [1, 2, 5, 10, 20, 50, 100, 200] :: Array Int Int

arr :: Array (Int, Int) Int
arr = array ((-200, 0), (200, 7)) [(id <:> f) (x, y) | x <- [-200..200], y <- [0..7]]

f (x, y) | x < 0 = 0
         | x == 0 || y == 0 = 1
         | otherwise = sum $ map (\ci -> arr ! (x - coins ! ci, ci)) [0..y]

euler031 :: IO String
euler031 = return $ show $ arr ! (200, 7)

File: rust/src/solutions/euler031.rs
pub fn euler031() -> String {
    let mut arr = vec![vec![0; 8]; 201];
    let coins = [1, 2, 5, 10, 20, 50, 100, 200];
    for i in 0..=200 {
        arr[i][0] = 1;
    }
    for j in 1..coins.len() {
        for i in 0..coins[j] {
            arr[i][j] = arr[i][j - 1];
        }
        for i in coins[j]..=200 {
            arr[i][j] = arr[i - coins[j]][j] + arr[i][j - 1];
        }
    }
    arr[200][7].to_string()
}
